File: tools/win-asset-collector/src/map.ts
import type { RawDisk, RawWmi } from "./collect";

const MAX = 50;

const LAPTOP_CHASSIS = new Set([8, 9, 10, 11, 12, 14, 31]);
const TABLET_CHASSIS = new Set([30, 32]);

export type MappedFields = {
	type: string;
	serialNumber: string;
	manufacturer: string;
	model: string;
	processor: string;
	os: string;
	memory: string;
	hdd: string;
	deviceName: string;
	ip: string;
};

function truncate(value: string, max: number): string {
	return value.trim().slice(0, max);
}

export function detectType(raw: RawWmi): string {
	if (raw.chassis.some((c) => LAPTOP_CHASSIS.has(c))) {
		return "Laptop";
	}
	if (raw.chassis.some((c) => TABLET_CHASSIS.has(c))) {
		return "Tablet";
	}
	if (raw.pcType === 2) {
		return "Laptop";
	}
	return "Desktop";
}

function formatMemory(bytes: string): string {
	const value = Number(bytes);
	if (!Number.isFinite(value) || value <= 0) {
		return "";
	}
	return `${Math.round(value / 1024 ** 3)} GB`;
}

function formatDisks(disks: RawDisk[]): string {
	return disks
		.slice(0, 2)
		.map((disk) => {
			const size = formatBytes(disk.size);
			const name = truncate(disk.model, 24);
			return [name, size].filter(Boolean).join(" ");
		})
		.join(" + ");
}

function formatBytes(value: string): string {
	const bytes = Number(value);
	if (!Number.isFinite(bytes) || bytes <= 0) {
		return "";
	}
	const units = ["GB", "TB"];
	let size = bytes / 1024 ** 3;
	let unit = units[0];
	if (size >= 1024) {
		size /= 1024;
		unit = units[1];
	}
	return `${size >= 100 ? Math.round(size) : size.toFixed(1)} ${unit}`;
}

function firstIp(ips: string[]): string {
	for (const ip of ips) {
		if (/^(\d{1,3}\.){3}\d{1,3}$/.test(ip)) {
			return ip;
		}
	}
	return "";
}

export function mapFields(raw: RawWmi): MappedFields {
	return {
		type: detectType(raw),
		serialNumber: truncate(raw.serial, MAX),
		manufacturer: truncate(raw.manufacturer, MAX),
		model: truncate(raw.model, MAX),
		processor: truncate(raw.processor, MAX),
		os: truncate(raw.os, MAX),
		memory: truncate(formatMemory(raw.memory), MAX),
		hdd: truncate(formatDisks(raw.disks), MAX),
		deviceName: truncate(raw.hostname, MAX),
		ip: truncate(firstIp(raw.ipv4), MAX),
	};
}

File: tools/go-win-asset-collector/internal/mapfields/map.go
// Package mapfields maps the raw platform snapshot into the asset form fields,
// equivalent to the original collector's src/map.ts so the payload contract
// with /app/assets?new= stays unchanged.
package mapfields

import (
	"math"
	"net"
	"strconv"
	"strings"

	"go-win-asset-collector/internal/collect"
)

// Fields is the payload sent to the ITSM app (the zod-whitelisted keys).
type Fields struct {
	Type         string `json:"type"`
	SerialNumber string `json:"serialNumber"`
	Manufacturer string `json:"manufacturer"`
	Model        string `json:"model"`
	Processor    string `json:"processor"`
	OS           string `json:"os"`
	Memory       string `json:"memory"`
	HDD          string `json:"hdd"`
	DeviceName   string `json:"deviceName"`
	IP           string `json:"ip"`
}

const max = 50

var laptopChassis = map[int]bool{8: true, 9: true, 10: true, 11: true, 12: true, 14: true, 31: true}
var tabletChassis = map[int]bool{30: true, 32: true}

// Map converts raw collected data into the form fields.
func Map(raw *collect.Raw) Fields {
	return Fields{
		Type:         DetectType(raw.Chassis, raw.PCType),
		SerialNumber: truncate(raw.Serial),
		Manufacturer: truncate(raw.Manufacturer),
		Model:        truncate(raw.Model),
		Processor:    truncate(raw.Processor),
		OS:           truncate(raw.OS),
		Memory:       truncate(formatMemory(raw.Memory)),
		HDD:          truncate(formatDisks(raw.Disks)),
		DeviceName:   truncate(raw.Hostname),
		IP:           truncate(firstIP(raw.IPv4)),
	}
}

// DetectType determines Desktop / Laptop / Tablet from the SMBIOS chassis type.
func DetectType(chassis []int, pcType int) string {
	for _, c := range chassis {
		if laptopChassis[c] {
			return "Laptop"
		}
	}
	for _, c := range chassis {
		if tabletChassis[c] {
			return "Tablet"
		}
	}
	if pcType == 2 {
		return "Laptop"
	}
	return "Desktop"
}

// truncate keeps at most max runes, matching the original 50-char clamp.
func truncate(value string) string {
	runes := []rune(value)
	if len(runes) > max {
		return string(runes[:max])
	}
	return value
}

func formatMemory(bytes string) string {
	value, err := strconv.ParseFloat(bytes, 64)
	if err != nil || value <= 0 {
		return ""
	}
	return strconv.FormatInt(int64(math.Round(value/(1024*1024*1024))), 10) + " GB"
}

func formatDisks(disks []collect.Disk) string {
	n := len(disks)
	if n > 2 {
		n = 2
	}
	parts := make([]string, 0, n)
	for _, disk := range disks[:n] {
		size := formatBytes(disk.Size)
		name := truncateRunes(disk.Model, 24)
		switch {
		case name == "":
			parts = append(parts, size)
		case size == "":
			parts = append(parts, name)
		default:
			parts = append(parts, name+" "+size)
		}
	}
	return strings.Join(parts, " + ")
}

func formatBytes(value string) string {
	bytes, err := strconv.ParseFloat(value, 64)
	if err != nil || bytes <= 0 {
		return ""
	}
	size := bytes / 1024 / 1024 / 1024
	unit := "GB"
	if size >= 1024 {
		size /= 1024
		unit = "TB"
	}
	if size >= 100 {
		return strconv.FormatInt(int64(math.Round(size)), 10) + " " + unit
	}
	return strconv.FormatFloat(size, 'f', 1, 64) + " " + unit
}

func firstIP(ips []string) string {
	for _, ip := range ips {
		if net.ParseIP(ip) != nil && net.ParseIP(ip).To4() != nil {
			return ip
		}
	}
	return ""
}

func truncateRunes(value string, maxRunes int) string {
	runes := []rune(value)
	if len(runes) > maxRunes {
		return string(runes[:maxRunes])
	}
	return value
}
